assets/js/components/Dashboard/PermanentConfiguration/Workflows/GlEink/ConfigureScreensPage.tsx

```diff
@@ -217,6 +217,106 @@ const ConfigurePlaceCard: ComponentType<ConfigurePlaceCardProps> = ({
     Object.keys(existingPendingScreens).length > 0 ||
     newScreens.length > 0;
 
+  const deleteExistingPendingRow = (
+    screenID: string,
+    screen: ScreenConfiguration
+  ) => {
+    setExistingPendingScreens((prevState) => {
+      const newState = { ...prevState };
+      newState[screenID].is_deleted = true;
+      return newState;
+    });
+
+    setUpdatedPendingScreens((prevState) => {
+      const newState = [...prevState];
+      const index = newState.findIndex(
+        (screen) => screen.screen_id === screenID
+      );
+
+      if (index === -1) {
+        newState.push({
+          screen_id: screenID,
+          is_deleted: true,
+          app_id: "gl_eink_v2",
+          app_params: screen.app_params,
+        });
+      } else {
+        newState[index].is_deleted = true;
```

Comment:
`[...prevState]` creates only a shallow copy of `prevState`, so this statement will mutate the original.

For example:
```ts
const listOfObjs = [{is_deleted: false}, {is_deleted: false}, {is_deleted: false}];
const shallowCopy = [...listOfObjs];
shallowCopy[1].is_deleted = true;

console.log(shallowCopy);
// [{is_deleted: false}, {is_deleted: true}, {is_deleted: false}]
​
console.log(listOfObjs);
// [{is_deleted: false}, {is_deleted: true}, {is_deleted: false}]
```

https://react.dev/reference/react/useState#setstate
> If you pass a function as nextState, it will be treated as an updater function. **It must be pure**, should take the pending state as its only argument, and should return the next state.

This callback can be made pure in the following way:
```ts
(prevState) => {
  const index = newState.findIndex((screen) => screen.screen_id === screenID);

  if (index === -1) {
    const newDeletedScreen = {
      screen_id: screenID,
      is_deleted: true,
      app_id: "gl_eink_v2",
      app_params: screen.app_params,
    };

    return [...prevState, newDeletedScreen];
  } else {
    return [
      ...prevState.slice(0, i),
      {...prevState[index], is_deleted: true},
      ...prevState.slice(i + 1)
    ];
  }
}
```

Most of the other updaters passed to `set*` functions below are also impure—please adjust them to avoid mutating the original state values.

---

Side note: I realize it's ridiculously cumbersome to do immutable updates like this in JS. I bet there's a nice library out there to make the process easier, will ask around if other teams are using something like that.

Reply:
Wow, very interesting. I definitely need to pay more attention to things like this. Will address these now.

Reply:
I think I tackled them all. Did a lot of poking around to make sure nothing changed functionally, and everything is acting the same as before. Please do flag any I may have missed.